File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    val maxVal = 101

    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode?
        = if (list1 == null && list2 == null) {
            null
        } else {
            var node1 = list1
            var node2 = list2
            var mergeList: ListNode? = null
            var mergeNode: ListNode? = null

            while (node1 != null || node2 != null) {
                val val1 = node1?.`val` ?: maxVal
                val val2 = node2?.`val` ?: maxVal

                val newNode = if (val1 <= val2) {
                    node1 = node1!!.next
                    ListNode(val1)
                } else {
                    node2 = node2!!.next
                    ListNode(val2)
                }

                if (mergeList == null) {
                    mergeList = newNode
                    mergeNode = newNode
                } else {
                    mergeNode!!.next = newNode
                    mergeNode = newNode
                }
            }

            mergeList
        }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(
        list1: Option<Box<ListNode>>,
        list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        match (list1, list2) {
            (Option::None, Option::None) => Option::None,
            (Option::Some(nb), Option::None) => Option::Some(Box::new(ListNode {
                val: nb.val,
                next: (Solution::merge_two_lists(nb.next, Option::None)),
            })),
            (Option::None, Option::Some(nb)) => Option::Some(Box::new(ListNode {
                val: nb.val,
                next: (Solution::merge_two_lists(Option::None, nb.next)),
            })),
            (Option::Some(nb1), Option::Some(nb2)) => {
                if nb1.val <= nb2.val {
                    Option::Some(Box::new(ListNode {
                        val: nb1.val,
                        next: (Solution::merge_two_lists(nb1.next, Option::Some(nb2))),
                    }))
                } else {
                    Option::Some(Box::new(ListNode {
                        val: nb2.val,
                        next: (Solution::merge_two_lists(Option::Some(nb1), nb2.next)),
                    }))
                }
            }
        }
    }
}
